### PR TITLE
Fix: remove unnecessary caching of SELECT DISTINCT query while using repository functions

### LINKED ISSUE #2721: Caching queries
**Issue type:**

[ ] question
[* ] bug report
[ ] feature request
[* ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[* ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[* ] `0.2.7` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
In my case: **Redis cache**
https://github.com/typeorm/typeorm/blob/master/docs/caching.md : 
"... clearing cached results when you insert a new user" - similar (by remove) it doesn't

**repository<entity>**
1. find -> [1,2,3]
2. remove 2
3. find -> [1,2,3]; should be: [1,3]

Conclusion:
remove method doesn't clear cache key where was some entity

**Edited**
I forgot about config:
`
    cache: {
        type: 'redis',
        alwaysEnabled: true,
        duration: parseInt(AppConfig.DB.CACHE_REDIS_DURATION, 10),
        options: {}
    },`
and this line for "alwaysEnabled": https://github.com/typeorm/typeorm/blob/master/src/connection/BaseConnectionOptions.ts#L125

### PR DESCRIPTION
**Pull request by vasu2652**
Fix: remove unnecessary caching of preselect query while using the repository functions.

### Description of change

When using repository functions like findOne or find or findAndCount with cache enabled the default behavior of TypeOrm is trying to cache the intended query with the provided identifier and also a preselect query with a SELECT DISTINCT to resolve the PrimaryColumn.

Due to this data invalidation is going a bit out of control and we are getting to see stale data for a period of time after mutation on the entity.

Current Behaviour: Return stale data even after invalidation using the provided identifier.

Expected Behaviour: Return updated data after invalidation using provided the identifier.

Closes: #2721, #5983, #4277

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #2721 #5983 #4277`
- [x] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change N/A
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: test/github-issues/2721/repository-caching.ts [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+        // insert new entity
+        const user4 = new User();
+        user4.firstName = "Bakhrom";
+        user4.lastName = "Brochik";
+        user4.isAdmin = true;
+        await connection.getRepository(User).save(user4);
+
+        // without cache it must return really how many there entities are
+        const users2 = await await connection.getRepository(User).find({
```
Comment (shrujalshah28) [inline test/github-issues/2721/repository-caching.ts:69]: Remove one `await`

Comment (vasu2652) [reply test/github-issues/2721/repository-caching.ts]: oops. thank you!! 😊

**Comment by imnotjames** [comment]
Won't this effectively make the caching useless?  The most expensive query is often the one to find the right PKs and order.  Looking those rows up again by PK is by comparison much faster.

What is the cache caching?

**Comment by vasu2652** [comment]
> Won't this effectively make the caching useless? The most expensive query is often the one to find the right PKs and order. Looking those rows up again by PK is by comparison much faster.
> 
> What is cache caching?

@imnotjames  Yes, I agree with you. But...
Why would we need this explicit query caching when the user already has granular control over when to cache and when to invalidate. 
Due to this on the data change or mutation, there is no way we can invalidate this cached query, which ends up showing the state data.

What additional advantage this query is giving when we are already caching at a query level? and also this happens without the consent of the user until someone actually looks into the cache server.

**Comment by imnotjames** [comment]
>  Why would we need this explicit query caching when the user already has granular control over when to cache and when to invalidate.

Except they don't have any sort of control over the cache because you're removing it in this PR.

> What additional advantage this query is giving when we are already caching at a query level? and also this happens without the consent of the user until someone actually looks into the cache server.

If you are caching the other part of the query you must cache this part.  That's how TypeORM works.

TypeORM does a first pass here to pull the primary keys then does a second pass to pull the data by primary key - which is usually incredibly fast.

However: If you don't want caching do not enable caching for the query.  If you do any sort of caching you will have stale data because the caching implementation built into TypeORM is simple at best.

**Pull request closed without merging**